openmc/mgxs/library.py

```diff
@@ -0,0 +1,1469 @@
+        xs_type: {'macro', 'micro'}
+            Store the macro or micro cross section in units of cm^-1 or barns.
+            Defaults to 'macro'.
+        row_column: {'inout', 'outin'}
+            Store scattering matrices indexed first by incoming group and
+            second by outgoing group ('inout'), or vice versa ('outin').
+            Defaults to 'inout'.
+        libver : {'earliest', 'latest'}
+            Compatibility mode for the HDF5 file. 'latest' will produce files
+            that are less backwards compatible but have performance benefits.
+
+        Raises
+        ------
+        ValueError
+            When this method is called before a statepoint has been loaded
+
+        See also
+        --------
+        MGXS.build_hdf5_store(filename, directory, xs_type)
+
+        """
+
+        if self.sp_filename is None:
+            msg = 'Unable to export multi-group cross section library ' \
+                  'since a statepoint has not yet been loaded'
+            raise ValueError(msg)
+
+        cv.check_type('filename', filename, str)
+        cv.check_type('directory', directory, str)
+
+        import h5py
+
+        # Make directory if it does not exist
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        # Add an attribute for the number of energy groups to the HDF5 file
+        full_filename = os.path.join(directory, filename)
+        full_filename = full_filename.replace(' ', '-')
+        f = h5py.File(full_filename, 'w', libver=libver)
+        f.attrs['# groups'] = self.num_groups
+        f.close()
+
+        # Export MGXS for each domain and mgxs type to an HDF5 file
+        for domain in self.domains:
+            for mgxs_type in self.mgxs_types:
+                mgxs = self.all_mgxs[domain.id][mgxs_type]
+
+                if subdomains == 'avg':
+                    mgxs = mgxs.get_subdomain_avg_xs()
+
+                mgxs.build_hdf5_store(filename, directory, xs_type=xs_type,
+                                      nuclides=nuclides, row_column=row_column)
+
+    def dump_to_file(self, filename='mgxs', directory='mgxs'):
+        """Store this Library object in a pickle binary file.
+
+        Parameters
+        ----------
+        filename : str
+            Filename for the pickle file. Defaults to 'mgxs'.
+        directory : str
+            Directory for the pickle file. Defaults to 'mgxs'.
+
+        See also
+        --------
+        Library.load_from_file(filename, directory)
+
+        """
+
+        cv.check_type('filename', filename, str)
+        cv.check_type('directory', directory, str)
+
+        # Make directory if it does not exist
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        full_filename = os.path.join(directory, filename + '.pkl')
+        full_filename = full_filename.replace(' ', '-')
+
+        # Load and return pickled Library object
+        pickle.dump(self, open(full_filename, 'wb'))
```

Comment:
This file is being opened and never closed.

Something like this should be better:


```
with open(full_filename, 'wb') as f:
    pickle.dump(self, f)
```